ops/lib/__init__.py

```diff
@@ -0,0 +1,201 @@
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+import os
+import re
+
+from ast import literal_eval
+from importlib.util import module_from_spec
+from importlib.machinery import ModuleSpec
+from pkgutil import get_importer
+from types import ModuleType
+from typing import Tuple, Dict, List, Iterator, Optional
+
+
+_libraries = {}  # type: Dict[Tuple[str,str], List[_Lib]]
+
+_libline_re = re.compile(r'''^LIB([A-Z]+)\s+=\s+([0-9]+|['"][a-zA-Z0-9_.-@]+['"])\s*$''')
+_libname_re = re.compile(r'''^[a-z][a-z0-9]+$''')
+
+# Not perfect, but should do for now.
+_libauthor_re = re.compile(r'''^[A-Za-z0-9_+.-]+@[a-z0-9_-]+(?:\.[a-z0-9_-]+)*\.[a-z]{2,3}$''')
+
+
+def use(name: str, api: int, author: str) -> ModuleType:
+    """Use a library from the ops libraries.
+
+    Args:
+        name: the name of the library requested.
+        api: the API version of the library.
+        author: the author of the library. If not given, requests the
+            one in the standard library.
+    Raises:
+        ImportError: if the library cannot be found.
+        TypeError: if the name, api, or author are the wrong type.
+        ValueError: if the name, api, or author are invalid.
+    """
+    if not isinstance(name, str):
+        raise TypeError("invalid library name: {!r} (must be a str)".format(name))
+    if not isinstance(author, str):
+        raise TypeError("invalid library author: {!r} (must be a str)".format(author))
+    if not isinstance(api, int):
+        raise TypeError("invalid library API: {!r} (must be an int)".format(api))
+    if api < 0:
+        raise ValueError('invalid library api: {} (must be ≥0)'.format(api))
+    if not _libname_re.match(name):
+        raise ValueError("invalid library name: {!r} (chars and digits only)".format(name))
+    if not _libauthor_re.match(author):
+        raise ValueError("invalid library author email: {!r}".format(author))
+
+    versions = _libraries.get((name, author), ())
+    for lib in versions:
+        if lib.api == api:
+            return lib.import_module()
+
+    others = ', '.join(str(lib.api) for lib in versions)
+    if others:
+        msg = 'cannot find "{}" from {} with API {} (have {})'.format(name, author, api, others)
+    else:
+        msg = 'cannot find library "{}" from {}'.format(name, author)
+
+    raise ImportError(msg, name=name)
+
+
+def autoimport():
+    _libraries.clear()
+    for spec in _find_all_specs(sys.path):
+        lib = _parse_lib(spec)
+        if lib is None:
+            continue
+
+        versions = _libraries.setdefault((lib.name, lib.author), [])
+        versions.append(lib)
+        versions.sort(reverse=True)
```

Comment:
What about storing here a dict per api? E.g.:
```
d = _libraries.setdefault((lib.name, lib.author), {})
d[lib.api] = lib
```

This would simplify a lot the code above in `use()`.

Furthermore, this could expose a not nice situation: what if we have more than one api version for the same name/author?

Reply:
We discussed this on IRC some more.

```
<Chipaca> facubatista: it can't be exactly as you show because then we'd risk overwriting an newer patch-level with an older one for the same api
<Chipaca> but it's an easy tweak
<facubatista> yes, (api, patch)
<Chipaca> but, but, not sure it's simpler. OT1H it's got a better behaviour in a pathologic case
<Chipaca> facubatista: nah, you don't want the patch in that dict, just the api
<facubatista> Chipaca, in any case, even if we keep the current structure, let's think about "repeated" stuff there
<Chipaca> and if d[api] < the_one_i_just_found, then replace it
<Chipaca> facubatista: ok let's do that
<facubatista> mmm... you still need to put 5.4 in that structure, even if you have 5.5, because the user may specify 5.4
<Chipaca> facubatista: no they mayn't
<Chipaca> facubatista: the user may only specify api, not patch
<facubatista> ah, right
<Chipaca> so we only need to keep one
<Chipaca> facubatista: that's we the list is kept sorted
<Chipaca> facubatista: but
<Chipaca> facubatista: the common case is going to be to have just one of these things
<Chipaca> two is already stretching it
<Chipaca> so the current approach seems fine
<facubatista> Chipaca, ok
<Chipaca> if it turns out we're wrong about this and need to change to a dict because, it's a private, invisible change
<Chipaca> so i'm happy as is